### PR TITLE
[SPARK-23464][MESOS] Fix mesos cluster scheduler options double-escaping

### PR DESCRIPTION
**Pull request by krcz**
## What changes were proposed in this pull request?

Don't enclose --conf option value with "", as they are already escaped by shellEscape
and additional wrapping cancels out this escaping, making the driver fail to start.

This reverts commit 9b377aa49f14af31f54164378d60e0fdea2142e5 [SPARK-18114].

## How was this patch tested?

Manual test with driver command logging added.

Author: Marcin Kurczych <marcin.kurczych@iqvia.com>

### CODE REVIEW

**Comment by susanxhuynh** [comment]
Thanks for the PR! It seems that the previous attempt to fix this (SPARK-18114) was wrong -- I'm not sure why we didn't catch the problem before, maybe lack of testing? @krcz My suggestion for this patch is to add a test, in order to prevent another regression in the future. I've written a unit test for this -- you could do something similar:  https://github.com/mesosphere/spark/commit/4812ba3d10264f6d22ec654fa16b5810d70c27a9 I will also do more testing with my own integration tests. cc @skonto

**Comment by krcz** [comment]
@susanxhuynh Thanks for the example! I have added similar test, covering more cases. After you or someone else reviews it, I'll rebase the pull request.

File: resource-managers/mesos/src/test/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterSchedulerSuite.scala [hunk 1]
```diff
@@ -199,6 +199,38 @@ class MesosClusterSchedulerSuite extends SparkFunSuite with LocalSparkContext wi
...
+      new MesosDriverDescription("d1", "jar", mem, cpu, true,
+        command,
+        Map("spark.mesos.executor.home" -> "test",
+          "spark.app.name" -> "test",
+          // no special characters, wrap only
+          "spark.driver.extraJavaOptions" ->
+            "-XX+PrintGC -Dparam1=val1 -Dparam2=val2",
+          // special characters, to be escaped
+          "spark.executor.extraJavaOptions" ->
+            """-Dparam1="value 1" -Dparam2=value\ 2 -Dpath=$PATH"""),
```
Comment (susanxhuynh) [inline resource-managers/mesos/src/test/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterSchedulerSuite.scala]: I'm not sure the second param (param2) would be parsed correctly if you actually ran the command. Doesn't there need to be quotes around the space? Have you tested it and checked if the executor gets the correct value for param2?

**Review by susanxhuynh — COMMENTED** [review]
Thanks, @krcz . Overall looks good ... I left one question about one of the test parameters. Also, have you verified in the driver / executors that they receive the correct Java system properties (with the "-D" options)? -- I was going to test that myself.

File: resource-managers/mesos/src/test/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterSchedulerSuite.scala (hunk 1 above)
Comment (krcz) [reply resource-managers/mesos/src/test/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterSchedulerSuite.scala]: @susanxhuynh 
I have checked it and it worked. There don't need to be quotes, as the space has been escaped. Backslash stops it from being interpreted as a boundary between arguments and makes it being understood as simple space in the value.

File: resource-managers/mesos/src/test/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterSchedulerSuite.scala [hunk 2]
```diff
@@ -199,6 +199,38 @@ class MesosClusterSchedulerSuite extends SparkFunSuite with LocalSparkContext wi
     })
   }
 
+  test("properly wraps and escapes parameters passed to driver command") {
```
Comment (skonto) [inline resource-managers/mesos/src/test/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterSchedulerSuite.scala]: Does this test fail with the old code?

Comment (krcz) [reply resource-managers/mesos/src/test/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterSchedulerSuite.scala]: It does.

```
[info] - properly wraps and escapes parameters passed to driver command *** FAILED *** (154 milliseconds)
[info]   "test/./bin/spark-submit --name test --master mesos://mesos://localhost:5050 --driver-cores 1.0 --driver-memory 1000M --class mainClass --conf "spark.app.name=test" --conf "spark.mesos.executor.home=tes
t" --conf "spark.executor.extraJavaOptions="-Dparam1=\"value 1\" -Dparam2=value\\ 2 -Dpath=\$PATH"" --conf "spark.driver.extraJavaOptions="-XX+PrintGC -Dparam1=val1 -Dparam2=val2"" ./jar arg" did not contain "--
conf spark.driver.extraJavaOptions="-XX+PrintGC -Dparam1=val1 -Dparam2=val2"" (MesosClusterSchedulerSuite.scala:227)
[info]   org.scalatest.exceptions.TestFailedException:
[info]   at org.scalatest.Assertions$class.newAssertionFailedException(Assertions.scala:528)
[info]   at org.scalatest.FunSuite.newAssertionFailedException(FunSuite.scala:1560)
[info]   at org.scalatest.Assertions$AssertionsHelper.macroAssert(Assertions.scala:501)
[info]   at org.apache.spark.scheduler.cluster.mesos.MesosClusterSchedulerSuite$$anonfun$16.apply(MesosClusterSchedulerSuite.scala:227)
[info]   at org.apache.spark.scheduler.cluster.mesos.MesosClusterSchedulerSuite$$anonfun$16.apply(MesosClusterSchedulerSuite.scala:202)
```

Comment (blue666man) [reply resource-managers/mesos/src/test/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterSchedulerSuite.scala]: This looks ready to merge; anything else needed @skonto or @susanxhuynh ?

Comment (adyatlov) [reply resource-managers/mesos/src/test/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterSchedulerSuite.scala]: That would be great to merge it if you think it's ready.

Comment (susanxhuynh) [reply resource-managers/mesos/src/test/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterSchedulerSuite.scala]: Sorry for the delay. I was going to test this in DC/OS and haven't gotten a chance to do so.

Comment (skonto) [reply resource-managers/mesos/src/test/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterSchedulerSuite.scala]: @susanxhuynh I can do a test for this.

**Comment by skonto** [comment]
@susanxhuynh I tested this. Here is what you get:

```
./bin/spark-submit --deploy-mode cluster --name test --master mesos://spark.marathon.mesos:13830 --driver-cores 1.0 --driver-memory 1000M --class org.apache.spark.examples.SparkPi --conf spark.executor.extraJavaOptions="-Dparam1=\"value 1\" -Dparam2=value\\ 2 -Dpath=\$PATH" --conf spark.mesos.containerizer=mesos  --conf spark.mesos.executor.docker.image=skonto/spark-test:test https://...jar 10000
```
At the driver side in DC/OS:

```
/usr/lib/jvm/jre1.8.0_152/bin/java -cp /opt/spark/dist/conf/:/opt/spark/dist/jars/*:/etc/hadoop/ -Dspark.mesos.driver.frameworkId=bddf96ce-45f6-45f7-ac83-5a3622cafc41-0030-driver-20180612154509-0007 -Xmx1000M org.apache.spark.deploy.SparkSubmit --master mesos://zk://master.mesos:2181/mesos --conf spark.driver.memory=1000M --conf spark.driver.cores=1.0 --conf spark.mesos.executor.docker.image=skonto/spark-test:test --conf spark.app.name=test --conf spark.mesos.containerizer=mesos --conf spark.executor.extraJavaOptions=-Dparam1="value 1" -Dparam2=value\ 2 -Dpath=$PATH --conf spark.driver.supervise=false --class org.apache.spark.examples.SparkPi --name test --driver-cores 1.0 /mnt/mesos/sandbox/....jar 10000
```

**Comment by krcz** [comment]
@skonto 
The command you have pasted as "At the driver side in DC/OS:" concerns me. It doesn't seem escaped properly.  Is it the command to be run by shell or is it just space-joined arguments list (as in Spark Dispatcher web UI), which would explain why it is not escaped?

**Comment by skonto** [comment]
@krcz which one? Btw both job run successfully although might not have been escaped properly.

**Comment by krcz** [comment]
@skonto The one directly below  "At the driver side in DC/OS:" in your comment. But if it's output of ps that's ok, as `ps` wouldn't be able to discern between `command "foo bar"` and `command foo bar`.

**Comment by skonto** [comment]
@krcz yes it should be ok. So I verified it:


Dispatcher passes the following command to mesos:
```18/06/13 10:30:45 INFO MesosClusterScheduler: Command for the driver (test):./bin/spark-submit --name test --master mesos://zk://master.mesos:2181/mesos --driver-cores 1.0 --driver-memory 1000M --class org.apache.spark.examples.SparkPi --conf spark.executor.extraJavaOptions="-Dexecutor.test.param1=\"value 1\" -Dexecutor.test.param2=value\\ 2 -Dexecutor.test.path=\$PATH" --conf spark.mesos.containerizer=mesos --conf spark.driver.supervise=false --conf spark.app.name=test --conf spark.driver.memory=1000M --conf
```
spark-class will run this at the executor side:
```
exec /usr/lib/jvm/jre1.8.0_152/bin/java -cp '/opt/spark/dist/conf/:/opt/spark/dist/jars/*:/etc/hadoop/' '-Dexecutor.test.param1=value 1' '-Dexecutor.test.param2=value 2' '-Dexecutor.test.path=$PATH' -Xmx1024m org.apache.spark.executor.CoarseGrainedExecutorBackend --driver-url spark://CoarseGrainedScheduler@ip-10-0-0-77.us-west-1.compute.internal:33887 --executor-id 9 --hostname 10.0.0.77 --cores 3 --app-id bddf96ce-45f6-45f7-ac83-5a3622cafc41-0032-driver-20180613103044-0001
```

**Comment by felixcheung** [comment]
ok, could you rebase this PR

**Comment by felixcheung** [comment]
ok to test

**Comment by felixcheung** [comment]
@mgummelt too

**Comment by krcz** [comment]
@felixb @skonto 
I've triend rebasing, but it looks that the issue has been meanwhile fixed independently in [other pull request](https://github.com/apache/spark/pull/21014)  while someone was working on [SPARK-23941 issue](https://issues.apache.org/jira/browse/SPARK-23941). It looks that there is no unit test there though, so I can modify his PR to incorporate just the test. Do you want that or should I just close the pull request?

**Comment by felixb** [comment]
@krcz 
@felixcheung <- different felix ;)

**Comment by felixcheung** [comment]
@krcz how is your PR different from 21014?

**Comment by krcz** [comment]
@felixcheung It fixes the same problem so in terms of implementation it is not very different. But when I created this one 21014 didn't exist.

The only difference is that this PR adds an unit test, which 21014 had not. That's why I'm asking if I should modify this PR to just add the test.

**Comment by felixcheung** [comment]
yap, that sounds like a good idea

**Comment by vanzin** [comment]
I'm closing this since there's been no activity. If you plan on adding the test, update your branch, and tag this with the other bug number (I duped this one to the bug with the fix attached to it).

**Pull request closed without merging**